[settings] Enhance edits components with dynamic label and help text handling

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/54826

This PR improves the rendering of settings fields, addressing issues related to labels, placeholders, and descriptions. It ensures text fields without a desc property are handled appropriately, respects the placeholder property, and shows the help text.

In classic settings, we render the help text as a tooltip. It's possible to do the same thing by modifying the edit components. However, since we want to reuse the upstream components in the future, at this moment, I don't want to customize the edit components too much. So I just pass the help text to the `help` prop to `*Control` components`. If we get the design requirement to render the help text as a tooltip, we can customize the edit components then but should also request the upstream components to support that.

Changes:

- Refactor checkbox, input, and select edit components to support optional help text
- Add utility function to determine label and help text based on setting configuration
- Update transformers to handle label and help text for various setting types

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add/Enable Gutenberg 19.8 or earlier
2. Turn on `settings` feature flag
3. `WooCommerce > Settings`
4. See that text inputs, checkboxes, and select inputs have labels and help text
5. Go to `Emails > Email sender options`
6. See `"From" name` and `"From" address` input fields don't have unexpected id labels 
7. See `Header image` input has placeholder text `N\A`

![Screenshot 2025-01-27 at 17 13 21](https://github.com/user-attachments/assets/6d2a1466-9b9e-4af4-bec0-5d8ea2e8f2fe)

![Screenshot 2025-01-27 at 17 12 50](https://github.com/user-attachments/assets/166e9703-e58e-42d8-aa20-f5288e4753ab)

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Enhance edits components with dynamic label and help text handling

- Refactor checkbox, input, and select edit components to support optional help text
- Add utility function to determine label and help text based on setting configuration
- Improve component flexibility by allowing more dynamic rendering of form controls
- Update transformers to handle label and help text for various setting types
- Modify test cases to reflect new component structure

chihsuan committed:
Add changelog

psealock reviewed (commented):
The `label` prop is now rendering the `id`. There is something internal to `*Control` components that use the `id` as the label when the label isn't present.

### Trunk
<img width="576" alt="image" src="https://github.com/user-attachments/assets/fe1060cd-ae10-4400-b64e-2fe6ccd7c6d3" />

### This branch
<img width="553" alt="image" src="https://github.com/user-attachments/assets/ea56454c-386d-4b2f-be83-34b7503f45b3" />

I think the boolean is supplied as a string, so something like this works. Alternatively, we should `Boolean()` the value to avoid this.

```js
const helpText =
		description === 'true' || description === 'false' ? '' : description;
```

chihsuan commented:
> The `label` prop is now rendering the `id`. There is something internal to `*Control` components that use the `id` as the label when the label isn't present.
> 
> ### Trunk
> <img alt="image" width="576" src="https://private-user-images.githubusercontent.com/1922453/407074157-fe1060cd-ae10-4400-b64e-2fe6ccd7c6d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwMzYzMTIsIm5iZiI6MTczODAzNjAxMiwicGF0aCI6Ii8xOTIyNDUzLzQwNzA3NDE1Ny1mZTEwNjBjZC1hZTEwLTQ0MDAtYjY0ZS0yZmU2Y2NkN2M2ZDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjhUMDM0NjUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTliZGMyYjU4NDZkMTEwNWJlOWQzYmNkZTMyZTEzZGZmNzg3Mjg3YWQ1MzcwOWVkOTVjMDc5ZWY2NjkxN2ExYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.PwyjDq3yG4caw2GTthDQABxacN-zYxx_-e7PjNEzpUI">
> ### This branch
> <img alt="image" width="553" src="https://private-user-images.githubusercontent.com/1922453/407074495-ea56454c-386d-4b2f-be83-34b7503f45b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwMzYzMTIsIm5iZiI6MTczODAzNjAxMiwicGF0aCI6Ii8xOTIyNDUzLzQwNzA3NDQ5NS1lYTU2NDU0Yy0zODZkLTRiMmYtYmU4My0zNGI3NTAzZjQ1YjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjhUMDM0NjUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjFlNmQwZGZmNjI2NWMwNDFhOTUxMDlkNzQzZTU0MjBkYzQ3YTg1NDNiMWM4NTY0MzQwNzRlOWI3OWFiNGQ5OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.1sku6DMp1BrRwAu1s0IZiL7Stl53oC_Wdv1e7FM1I20">
> I think the boolean is supplied as a string, so something like this works. Alternatively, we should `Boolean()` the value to avoid this.
> 

@psealock This is weird. I'm not seeing the same behavior on my end. The following is what I'm seeing on this branch:

![Screenshot 2025-01-28 at 11 55 31](https://github.com/user-attachments/assets/e8f7345c-c508-4630-b118-9db92c19d1c8)


It should already be handled here:

https://github.com/woocommerce/woocommerce/blob/41ed7cad2cf6ca31c464be5b5d4100f07161dfc6/packages/js/settings-editor/src/components/inputEdit/index.tsx#L19-L22

Could you check if you're still seeing the issue? I can see that select controls are still using the `id` as the label though. I'll fix that.

chihsuan committed:
Fix id as label issue for select and checkbox

ilyasfoo commented on packages/js/settings-editor/src/components/checkbox-edit/index.tsx:
```diff
@@ -10,25 +10,28 @@ import type { DataFormControlProps } from '@wordpress/dataviews';
-	const { id, getValue, label } = field;
-	const value = getValue( { item: data } );
+export const getCheckboxEdit =
+	( help?: React.ReactNode ) =>
```
A few questions:

1. I'm just curious, is this shape an established pattern you're following?
2. Instead of committing to positional arguments, would it be better if we use named parameter instead? I think it's better for the long term in case we expect the list of parameters to grow.

chihsuan replied:
> 1. I'm just curious, is this shape an established pattern you're following?

Nope,  this is a temporary workaround. Ideally, we should not have to do this or even create our own Edits components but retrieve the required data from the fields, however, Gutenberg doesn't support this. We should contribute upstream to make this better.

> Instead of committing to positional arguments, would it be better if we use named parameter instead? I think it's better for the long term in case we expect the list of parameters to grow.

I don't really expect the list of parameters to grow. I think we can use positional arguments first and then convert to named parameters when the number of arguments grows. Does that sound good?

ilyasfoo replied:
Sounds good!

ilyasfoo reviewed (commented):
@chihsuan Testing well, but a few questions

ilyasfoo reviewed (approved):
Thanks, LGTM

psealock commented on packages/js/settings-editor/src/hooks/utils/transformers.tsx:
```diff
@@ -11,12 +11,25 @@ import type { Field, FormField } from '@wordpress/dataviews';
+ * Helper function to determine label and help text from setting
+ * If desc_tip is true, we should display the desc as the help text.
+ */
+const getLabelAndHelp = (
```
👍🏽 For reusable function - text areas and surely a few others will make use as well

psealock commented on packages/js/settings-editor/src/hooks/utils/transformers.tsx:
```diff
@@ -11,12 +11,25 @@ import type { Field, FormField } from '@wordpress/dataviews';
+		label: setting.desc_tip === true ? '' : setting.desc,
+		help: setting.desc_tip === true ? setting.desc : setting.desc_tip,
+	};
+};
```
To make this function reusable it looks like the description can be `settings.desc || setting.description`.

https://github.com/woocommerce/woocommerce/blob/d05829296d382c8251db90ee0603b86cb9b295a9/plugins/woocommerce/includes/gateways/paypal/includes/settings-paypal.php#L122-L129

Also, should we explicitly handle a `false` value for `desc_tip`? I haven't found an example in our codebase besides some defaults, but it could exist out there.

chihsuan replied:
Great suggestions. I've improved the function and added test cases for it. cf9e65c94b6bc1ab05fb852736a98e1a25f36af9

psealock reviewed (commented):
Thanks @chihsuan, this is working as expected. I'm not sure what happened. Just a few comments for improving `getLabelAndHelp`.

<img width="695" alt="image" src="https://github.com/user-attachments/assets/0864acdf-f6fa-4099-97a0-318ab88bb44d" />

chihsuan committed:
Improve getLabelAndHelp utility function with more robust description handling

- Enhance label and help text extraction from settings
- Add support for both 'desc' and 'description' fields
- Improve handling of desc_tip scenarios
- Add comprehensive test cases for different input scenarios
- Update TypeScript type definitions to include optional 'description' field

chihsuan committed:
Add docs

psealock reviewed (approved):
Thanks for the changes, looking great 🚢

psealock commented:
@chihsuan if you don't mind, I will merge this now with the updates. I'd like to use `getLabelAndHelp` in the textarea implementation as well as the upcoming radio and multiselect.
